### PR TITLE
Emit a single error on if expr with expectation and no else clause

### LINKED ISSUE #60254: Deduplicate diagnostics in `if-without-else-as-fn-expr.rs`
https://github.com/rust-lang/rust/pull/59288 introduces some duplicates in diagnostics in `if-without-else-as-fn-expr.rs`. Specifically, you have e.g.:
```
error[E0308]: mismatched types
  --> $DIR/if-without-else-as-fn-expr.rs:35:20
   |
LL |       let x: usize = if let 0 = 1 {
   |  ____________________^
LL | |         return 3;
LL | |     };
   | |_____^ expected usize, found ()
   |
   = note: expected type `usize`
              found type `()`

error[E0317]: if may be missing an else clause
  --> $DIR/if-without-else-as-fn-expr.rs:35:20
   |
LL |       let x: usize = if let 0 = 1 {
   |  _________-__________^
   | |         |
   | |         expected because of this assignment
LL | |         return 3;
LL | |     };
   | |_____^ expected usize, found ()
   |
   = note: expected type `usize`
              found type `()`
   = note: `if` expressions without `else` evaluate to `()`
   = help: consider adding an `else` block that evaluates to the expected type
```
We want to get rid of the first one since the second is better...

...but this is not done in https://github.com/rust-lang/rust/pull/59288 since I wanted to avoid too many temporary changes for diagnostics that may get obsoleted quickly.

### PR DESCRIPTION
**Pull request by estebank**
Fix https://github.com/rust-lang/rust/issues/60254.

r? @Centril

### CODE REVIEW

File: src/librustc_typeck/check/_match.rs [hunk 1]
```diff
@@ -185,11 +193,12 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         span: Span,
         then_expr: &'tcx hir::Expr,
         coercion: &mut CoerceMany<'tcx, '_, rustc::hir::Arm>,
-    ) {
+    ) -> bool {
```
Comment (Centril) [inline src/librustc_typeck/check/_match.rs]: Add to the doc comment:
```
/// Returns `true` if there was an error forcing the coercion to the `()` type.
```

File: src/librustc_typeck/check/_match.rs [hunk 2]
```diff
@@ -112,19 +112,27 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
-            all_arms_diverge &= self.diverges.get();
-
-            let span = expr.span;
-
+            let arm_ty = if source_if && if_no_else && i != 0 && self.if_fallback_coercion(
+                expr.span,
+                &arms[0].body,
+                &mut coercion,
+            ) {
+                    tcx.types.err
```
Comment (Centril) [inline src/librustc_typeck/check/_match.rs]: Indentation is off here.

File: src/librustc_typeck/check/_match.rs [hunk 3]
```diff
@@ -112,19 +112,27 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+            } else {
+                // Only call this if this is not an `if` expr with an expected type and no `else`
+                // clause to avoid duplicated type errors. (#60254)
+                let arm_ty = self.check_expr_with_expectation(&arm.body, expected);
+                all_arms_diverge &= self.diverges.get();
```
Comment (Centril) [inline src/librustc_typeck/check/_match.rs]: What is the reason for making this diverge bit conditional?

**Comment by estebank** [comment]
```
Dist cargo stage1 (mips64el-unknown-linux-gnuabi64)
Building stage1 tool cargo (mips64el-unknown-linux-gnuabi64)
 Downloading‌ crates ...‌
error:‌ failed to download from `https:&#x2F;&#x2F;crates.io&#x2F;api&#x2F;v1&#x2F;crates&#x2F;openssl-src&#x2F;111.3.0+1.1.1c&#x2F;download`‌

Caused by:
  [35] SSL connect error (error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure)
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "mips64el-unknown-linux-gnuabi64" "-Zdual-proc-macros" "-Zbinary-dep-depinfo" "-j" "2" "--release" "--locked" "--color" "always" "--manifest-path" "/checkout/src/tools/cargo/Cargo.toml" "--features" "rustc-workspace-hack/all-static" "--message-format" "json-render-diagnostics"
expected success, got: exit code: 101
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap dist --host mips64el-unknown-linux-gnuabi64 --target mips64el-unknown-linux-gnuabi64
Build completed unsuccessfully in 1:18:54
```

@bors retry

**Pull request merged**